Feature/form summary footer

SonyN-nav opened the pull request:
### Description

<!-- PR description/motivation, summary of changes, links to issue/slack discussion etc. -->

### Component Checklist 📝

- [x] JSDoc
- [x] Examples
    - website examples should be the only remaining
- [x] Documentation / Decision Records
- [ ] Storybook
- [ ] Style mappings (`@navikt/core/css/config/_mappings.js`)
- [x] Component tokens (`@navikt/core/css/tokens.json`)
- [x] CSS class deprecations (`@navikt/aksel-stylelint/src/deprecations.ts`)
- [ ] Exports (`@navikt/core/react/src/index.ts` and `@navikt/core/react/package.json`)
- [x] New component? CSS import (`@navikt/core/css/index.css`)
- [x] Breaking change? Update migration guide. Consider codemod.
- [ ] Changeset (Format: `<Component>: <gitmoji?> <Text>.` E.g. "Button: :sparkles: Add feature xyz.")

SonyN-nav committed:
✨ Adds FormSummary footer component

Adds a `FormSummaryFooter` component to the `FormSummary`.

Moves the `EditLink` from the header to the new footer for improved structure and consistency.

SonyN-nav committed:
🚨 Adds usage warning for incorrect component placement

Adds a warning to prevent incorrect placement of a component within another.

This improves DX by guiding developers to use the intended structure for form summary components.

KenAJoh commented on @navikt/core/react/src/form/form-summary/FormSummary.tsx:
Can also remove Editlink from within Header in JsDoc 📝

KenAJoh commented on @navikt/core/react/src/form/form-summary/FormSummaryFooter.tsx:
```diff
@@ -0,0 +1,21 @@
+  HTMLElement,
+  FormSummaryFooterProps
+>(({ children, className, ...rest }, ref) => (
+  <footer
```
Based on this issue #3951, i think we can get away with just using a `div` here.

SonyN-nav replied:
So rather than footer tag, we do a div wrapping? Reading the linked issue, should the header be rewritten to be a div as well? Perhaps preferable to do in a separate task.

KenAJoh replied:
Yes we can just use a regular `div` here, and update the type to use `HTMLDIvElement` 👍 
As for the header-change, that can probably be a separate task yes ✅

KenAJoh commented on @navikt/core/react/src/form/form-summary/FormSummaryFooter.tsx:
```diff
@@ -0,0 +1,21 @@
+  <footer
+    ref={ref}
+    {...rest}
+    className={`navds-form-summary__footer${className ? ` ${className}` : ""}`}
```
We normally use the package `clsx` to handle className collections. Since we currently support two versions of our CSS-package, we have wrapped this within the hook 
```
const { cn } = useRenameCSS();
```

```
    className={cn("navds-form-summary__footer", className)}

```

KenAJoh commented on @navikt/core/css/form/form-summary.css:
`form-summary.darkside.css` also needs to be updated 📝

KenAJoh commented on @navikt/core/react/src/util/usageWarnings.ts:
```diff
@@ -0,0 +1,36 @@
+  severity: "warn" | "error" = "warn",
+) {
+  React.useEffect(() => {
+    const hasType = React.Children.toArray(children).some(
```
One disadvantage of `React.Children.toArray` is that we only check one level now, so different layouts or compositions might break. One idea is to create a Provider/hook combo that pretty much does the exact same, but ignoring nesting-depth or compositions. (Just mocked this together, so unsure if its even valid code)
```tsx
// Header
<ComponentWarningBoundary componentName="FormSummary.EditLink">
...
</ComponentWarningBoundary>


// Editlink
useComponentWarningBoundary("FormSummary.EditLink");

// util..
function useComponentWarningBoundary(componentName: string) {
  const context = useContext(ComponentWarningContext);

  useEffect(() => {
    if(componentName ==! context?.componentName){ 
      return;
    }
    issueWarning(...);
  },[context])
}
```

SonyN-nav replied:
Agree, still massaging this util. Exploring if there are alternate approaches, the less you have to add to the component it warns about the better I think.

SonyN-nav replied:
New variant pushed

KenAJoh commented on @navikt/core/react/src/form/form-summary/FormSummary.tsx:
`@navikt/core/react/src/form/form-summary/index.ts` needs to be updated with Footer export 🏷️

KenAJoh reviewed (changes requested):
Unsure if PR was ready for review since it in `draft`-state so just ignore my comments if not for now 🙈

SonyN-nav committed:
✨ Adds edit link to FormSummaryHeader

Adds an edit link to the FormSummaryHeader component during testing.

This allows users to easily navigate to the corresponding form section for editing.

Removes requirement for FormSummaryHeader to include EditLink - now optional

SonyN-nav committed:
✨ Moves EditLink to FormSummary.Footer

Moves `EditLink` from `FormSummary.Header` to `FormSummary.Footer` for better component structure and to resolve design issues.

Adds `FormSummaryFooter` component and styles to support the `EditLink`.

Implements `useWarnIfContainsAny` hook to provide development-only warnings for incorrect usage of `FormSummary` subcomponents.

SonyN-nav committed:
Merge branch 'main' into feature/form-summary-footer

KenAJoh commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,20 +1,56 @@
+ * ❌ Do not place `<FormSummary.EditLink>` here — it belongs in `<FormSummary.Footer>`.
+ * ❌ Do not place `<FormSummary.Answers>` here - it belongs under <FormSummary>.
+ *
+ * ✔ See documentation: https://aksel.nav.no/komponenter/core/formsummary
```
JsDoc supports markdown, so this link can be collapsed

SonyN-nav committed:
Merge branch 'main' into feature/form-summary-footer

SonyN-nav committed:
🏗️  Refactors FormSummary structure and adds slot warnings

Refactors the FormSummary component to improve structure and provide better guidance for developers.

- Introduces `FormSummary.Header` and `FormSummary.Footer` components as dedicated slots, enhancing the component's API.
- Implements `useForbidSlot` and `useRequireSlot` hooks to provide development-only warnings when components are misused within the wrong FormSummary slots, improving component usability and preventing common integration issues.
- Removes the previous complex logic for warnings within `FormSummaryHeader` in favor of the new slot-based approach.

✅ Renames usageWarnings to slotWarningProvider

🚚 Moves and renames `usageWarnings.tsx` to `slotWarningProvider.tsx` for clarity.

This change improves the naming and location of the slot warning utilities,
making their purpose more explicit and easier to discover.
It also enhances the documentation and debug tooling.

KenAJoh committed:
:recycle: Updated API for slotwarning

KenAJoh committed:
:memo: Added sb-story for testing

KenAJoh committed:
:lipstick: Remove extra border-radius in footer

KenAJoh committed:
:memo: Updated examples

KenAJoh committed:
:memo: Added changeset

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummary.stories.tsx:
```diff
@@ -427,3 +439,34 @@ export const Chromatic: StoryObj<typeof FormSummary> = {
+);
+
+_DevWarning_MisuseInHeaderExample.parameters = {
+  docs: { disable: true },
```
What does this do?

KenAJoh replied:
Nothing it seems, only useful if we have autodocs enabled

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummary.stories.tsx:
```diff
@@ -312,6 +317,9 @@ export const Empty: StoryFn<typeof FormSummary> = () => (
+      </FormSummary.Footer>
     </FormSummary>
   </VStack>
 );
```
The "Empty" story has multiple examples that has not been updated. But we could discuss the need for these.

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryEditLink.tsx:
```diff
@@ -24,17 +16,23 @@ export const FormSummaryEditLink: OverridableComponent<
-      className={cn("navds-form-summary__edit", className)}
+    <CompositionWarning.Forbidden
+      name="FormSummary.Header"
+      message="<FormSummary.EditLink> should not be placed in <FormSummary.Header>. See: https://aksel.nav.no/komponenter/core/formsummary"
```
I suggest emphasizing that it hasn't always been like this:

```suggestion
      message="<FormSummary.EditLink> should not be placed in <FormSummary.Header> anymore. See https://aksel.nav.no/komponenter/core/formsummary"
```

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryEditLink.tsx:
```diff
@@ -1,21 +1,13 @@
-   *
-   * [How to write good links](https://aksel.nav.no/god-praksis/artikler/lenker).
-   * @default "Endre svar"
-   */
```
I get that we probably don't need an explenation for these, but wouldn't it be good to keep the `@default` part?

KenAJoh replied:
👍 Re-added them

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,28 +1,31 @@
-      {children}
-    </header>
+    <CompositionWarning.Root name="FormSummary.Header">
+      <div
```
Consider including the change from `<header>` to `<div>` in the changeset.

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,28 +1,31 @@
 
 export const FormSummaryHeader = forwardRef<
-  HTMLDivElement,
+  HTMLHeadingElement,
```
I believe this was correct before

```suggestion
  HTMLDivElement,
```

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,28 +1,31 @@
-  /**
-   * Must include `<FormSummary.Heading>` and optionally `<FormSummary.EditLink>`.
-   */
   children: React.ReactNode;
```
Why remove the jsdoc? Might not show up in the IDE, but shows up in the prop list on the website.

Should at least be consistent with the other components.

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummary.tsx:
```diff
@@ -53,14 +58,16 @@ export interface FormSummaryProps extends HTMLAttributes<HTMLDivElement> {
+   *   </FormSummary.Footer>
    * </FormSummary>
    */
   children: React.ReactNode;
```
Probably not relevant to include an example here (?)

KenAJoh replied:
Agreed

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryFooter.tsx:
```diff
@@ -0,0 +1,30 @@
+ */
+export interface FormSummaryFooterProps
+  extends React.HTMLAttributes<HTMLDivElement> {
+  children: React.ReactNode;
```
See comment about jsdoc in FormSummaryHeader

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryFooter.tsx:
```diff
@@ -0,0 +1,30 @@
+  children: React.ReactNode;
+}
+
+export const FormSummaryFooter = forwardRef<
```
Did you consider merging Footer and EditLink? Could look something like this:
```tsx
<div data-color="info" className="navds-form-summary__footer">
    <Link
        ref={ref}
        as={as}
        {...rest}
        className={cn("navds-form-summary__edit", className)}
    >
        {children || translate("editAnswer")}
    </Link>
</div>
```
The downside is that you wouldn't be able to set attributes on the div.

This has probably already been discussed, but I'm curious about the arguments 😄 Do we want people to be able to put other things in the footer?

KenAJoh replied:
For me it's more about the direction we want to take the system API, where i would prefer more composition instead of inheritance in the long run. While this can lead to "harder to updated" changes like this move of editlink -> footer, nothing actually breaks anywhere. 

Right now we don't have any concrete examples of other elements in the Footer, but if integrate the link into it we are kind of stuck. If a team comes with a great idea, we might need to break all use of Footer to accommodate this. One recent case of this is `Fileupload.Item` where some wanted custom actions/buttons. We basically had to implement a new API just to avoid introducing breaking changes. Over time we end up with varying APIs across the system.

HalvorHaugan replied:
I don't necessarily disagree, but I worry it will become challenging to maintain consistent UI across products (i.e. how to force product teams to update their code after changes like this).

KenAJoh replied:
That's true, i think the `CompositionWarning` API will help quite a bit to push people to upgrade. We could consider adding on to it to increase the disruption when needed. 

**Unfinished idea: **
1. We can tag something as forbidden for a specific version
- If they are on same major: console.warn in dev
- If they are on next major: Throw Error in dev

2. We can create a Eslint-plugin that caches those cases (a lot of work 😓 )

**Not actually a serious idea**
- If they are on next major, change styling to a flashing red box (or something awful) in dev

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,28 +1,31 @@
   children: React.ReactNode;
 }
 
 export const FormSummaryHeader = forwardRef<
```
I see that we forgot to put jsdoc on the components themselves (for when you import the subcomponents separately instead of using the dot notation). We Don't need to fix it in this PR though.

HalvorHaugan commented on aksel.nav.no/website/pages/eksempler/formsummary/demo.tsx:
```diff
@@ -62,6 +61,9 @@ const Example = () => {
           <FormSummary.Value>SMS</FormSummary.Value>
         </FormSummary.Answer>
       </FormSummary.Answers>
+      <FormSummary.Footer>
```
My OCD really wants an empty line before the footer for consistency 😁

```suggestion

      <FormSummary.Footer>
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tsx:
```diff
@@ -0,0 +1,77 @@
+ *
+ * Usage:
+ * - Wrap slot components with <CompositionWarning.Root name="FormSummary.Header">...</CompositionWarning.Root>
+ * - In child: `<CompositionWarningForbidden.Forbidden name="FormSummary.Header" />` to forbid slot.
```
```suggestion
 * - In child: `<CompositionWarning.Forbidden name="FormSummary.Header" />` to forbid slot.
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tsx:
```diff
@@ -0,0 +1,77 @@
+/**
+ * Enables warnings only in development or when forcedDebug is set.
+ */
+const isDev = forcedDebug || process.env.NODE_ENV === "development";
```
Why not just `const isDev = process.env.NODE_ENV !== "production"` ?

KenAJoh replied:
Agreed, changed to !== production

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tsx:
```diff
@@ -0,0 +1,77 @@
+type CompositionWarningForbiddenProps = {
+  children?: React.ReactElement;
+  /**
+   * Name of the composition that is forbidden.
```
I find this description confusing. What about something like this?

```suggestion
   * Name of the parent slot component where the child is not allowed.
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tsx:
```diff
@@ -0,0 +1,77 @@
+const isDev = forcedDebug || process.env.NODE_ENV === "development";
+
+type CompositionWarningContextType = {
+  name: CompositionName;
```
Maybe we should have jsdoc here as well?

```suggestion
  /**
   * Name of the slot component we want to check.
   */
  name: CompositionName;
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tsx:
```diff
@@ -0,0 +1,77 @@
+   */
+  name: CompositionName;
+  /**
+   * Warning message to display if invalid.
```
```suggestion
   * Warning message to display if the child is found.
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tests.stories.tsx:
```diff
@@ -0,0 +1,120 @@
+      if (args.length > 0) {
+        setLastMessage(String(args[0]));
+      }
+      return original.apply(console, args as unknown as [any]);
```
Seems to work without the type casting
```suggestion
      return original.apply(console, args);
```

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tests.stories.tsx:
```diff
@@ -0,0 +1,120 @@
+      <div data-testid="last-message" style={{ display: "none" }}>
+        {lastMessage}
+      </div>
+      {children}
```
I think this could be visible. Would just be nice.

KenAJoh replied:
What do you mean?

HalvorHaugan replied:
That I would consider removing `style={{ display: "none" }}`

KenAJoh replied:
Ofc, yea can remove those 👍

HalvorHaugan commented on @navikt/core/css/darkside/form/form-summary.darkside.css:
```diff
@@ -81,3 +81,13 @@
+  display: flex;
+  align-items: flex-start;
+  gap: var(--ax-space-8);
+  flex: 1 0 0;
```
I believe these are redundant (display, align-items, gap, flex)

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryEditLink.tsx:
```diff
@@ -24,17 +16,23 @@ export const FormSummaryEditLink: OverridableComponent<
+        {...rest}
+        className={cn("navds-form-summary__edit", className)}
+      >
+        {children || translate("editAnswer")}
```
We decided to add a pencil icon, but Sjur forgot to send us the updated Figma link: https://www.figma.com/design/303AiTt7dkC8BihRuCIGhQ/Feat--FormSummary-flytte-link?node-id=7417-49999&t=0aOQXmceKaBn7e8B-4

We also discussed having/recommending unique links (e.g. "Endre svar på personalia"), but I think we should wait and see how this works out first. (We can huddle about it if you want.)

KenAJoh replied:
Added now 🏁

KenAJoh committed:
:recycle: PR adjustments

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:tada: Added pencil-icon

KenAJoh committed:
:memo: Updated jsdoc

KenAJoh committed:
:memo: New changeset

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:fire: Removed forced-debug option

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
Apply suggestion from @HalvorHaugan

Co-authored-by: Halvor Haugan <83693529+HalvorHaugan@users.noreply.github.com>

KenAJoh committed:
:memo: show hidden warning-messages

HalvorHaugan commented on @navikt/core/css/darkside/form/form-summary.darkside.css:
```diff
@@ -22,7 +22,6 @@
 
 .aksel-form-summary__edit {
```
Ooh, we might want to do this
```suggestion
.aksel-form-summary__header .aksel-form-summary__edit {
```
(Then we know we won't break anything in the footer when removing it later.)

KenAJoh replied:
Done 🏁

HalvorHaugan commented on @navikt/core/react/src/form/form-summary/FormSummaryHeader.tsx:
```diff
@@ -1,28 +1,34 @@
+import { CompositionWarning } from "../../util/composition-warning";
 
+/**
+ * Header slot for section heading and actions in `FormSummary`.
```
"actions"?

KenAJoh replied:
Removed it, no need for jsdoc on the types.

HalvorHaugan commented on @navikt/core/react/src/util/composition-warning/CompositionWarning.tests.stories.tsx:
```diff
@@ -0,0 +1,116 @@
+
+  return (
+    <div>
+      <div data-testid="warn-count">{`Console warnings: ${count}`}</div>
+      <div data-testid="last-message">{`Laste msg: ${lastMessage}`}</div>
```
The tests are failing now 😅

KenAJoh replied:
Fixed 🙈

KenAJoh committed:
:memo: Scope edit-link css

KenAJoh committed:
:test_tube: Fixed tests

KenAJoh committed:
:memo: Sync EditLink updates in templates

HalvorHaugan commented on aksel.nav.no/website/pages/templates/soknad-oppsummeringsside-real-examples/aap.tsx:
```diff
@@ -110,6 +109,9 @@ function Example() {
+            </FormSummary.Footer>
           </FormSummary>
 
           <FormSummary>
```
You skipped one 😅 Line 120 ("Startdato")

KenAJoh committed:
:bug: Update last missing doc-change

HalvorHaugan commented on @navikt/core/css/darkside/form/form-summary.darkside.css:
```diff
@@ -81,3 +80,10 @@
+
+.aksel-form-summary__footer {
+  display: flex;
+  padding: var(--ax-space-16) var(--ax-space-20);
```
Responsive padding (as mentioned, in case you forgot)

KenAJoh replied:
Fixed 🚀

KenAJoh committed:
:wheelchair: Responsive padding
